### DISCUSSION TITLE
PyCharm Type check warning for FastAPI app object with uvicorn.run(app)

### DISCUSSION THREAD

**Discussion by eshfield** [fastapi/fastapi#8832] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def root():
    a = "a"
    b = "b" + a
    return {"hello world": b}


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description

**The problem:**
PyCharm gives ugly type check warning when I try to run FastAPI app with uvicorn programmatically passing to `uvicorn.run()` the application instance of the FastAPI class.

**Steps to reproduce:**
* Copy and paste to PyCharm the example code from the official FastAPI documentation [Debugging → Call uvicorn](https://fastapi.tiangolo.com/tutorial/debugging/?h=uvicorn#call-uvicorn)
* PyCharm will highlight the `app` argument at the last line `uvicorn.run(app, host="0.0.0.0", port=8000)` with warning: 
`Expected type 'Union[Type[ASGI2Protocol], (Union[HTTPScope, WebSocketScope, LifespanScope], () -> Awaitable[Union[HTTPRequestEvent, HTTPDisconnectEvent, WebSocketConnectEvent, WebSocketReceiveEvent, WebSocketDisconnectEvent, LifespanStartupEvent, LifespanShutdownEvent]], (Union[HTTPResponseStartEvent, HTTPResponseBodyEvent, HTTPServerPushEvent, HTTPDisconnectEvent, WebSocketAcceptEvent, WebSocketSendEvent, WebSocketResponseStartEvent, WebSocketResponseBodyEvent, WebSocketCloseEvent, LifespanStartupCompleteEvent, LifespanStartupFailedEvent, LifespanShutdownCompleteEvent, LifespanShutdownFailedEvent]) -> Awaitable[None]) -> Awaitable[None], str]', got 'FastAPI' instead`

I know that I can run uvicorn with the import string style (`uvicorn.run("main:app")` instead of `uvicorn.run(app)`) without any PyCharm warnings, but I need too pass exactly the app object, not the string.

And I know that I can suppress this PyCharm warning with `# noinspection PyTypeChecker` comment, but FastAPI is all about type hints and IDE-friendly way, so I guess that this kind of ugly workaround is not a good solution.

### Operating System

Windows

### Operating System Details

Windows 10 Home 21H1

### FastAPI Version

0.68.1

### Python Version

Python 3.9.7

### Additional Context

PyCharm Professional version: 2021.2.2

**Comment by STeveShary** [comment]
@DemX86 The issue is that Uvicorn has recently added typing for it's configuration and it wants an: ASGIApplication, Callable, or str.   This change was more recently added: https://github.com/br3ndonland/uvicorn/blob/5d59bcc39940567c2940aa52093fe051468b6046/uvicorn/config.py#L139

FastAPI extends from Starlette: https://github.com/tiangolo/fastapi/blob/master/fastapi/applications.py#L31 and the issue here is that Starlette is not typed to an `ASGIApplication`: https://github.com/encode/starlette/blob/master/starlette/applications.py#L12

For now, `# noinspection PyTypeChecker` is your best bet.

**Comment by Kludex** [comment]
It's more or less how @STeveShary said, but the issue is the annotation Starlette has on the `__call__` method (https://github.com/encode/starlette/blob/85316543e7aa69461d3bec536703e5ee6d432a40/starlette/applications.py#L110-L112).

`Scope`, `Receive` and `Send` are from Starlette itself and differ from `asgiref`.

In any case, the fix should be made in Starlette.

**Comment by mbwmbw1337** [comment]
Any updates getting this one fixed?

**Comment by Kludex** [comment]
No.

**Comment by DougPlumley** [comment]
I'm seeing this same error/warning in VS Code as well with the Python extension and Pylance.
